File: src/main/java/pl/marcinsalamandra/microbloggingapp/service/PostService.java
package pl.marcinsalamandra.microbloggingapp.service;

import lombok.RequiredArgsConstructor;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.Pageable;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;
import org.springframework.web.multipart.MultipartFile;
import pl.marcinsalamandra.microbloggingapp.dto.request.CreatePostDTO;
import pl.marcinsalamandra.microbloggingapp.dto.request.UpdatePostDTO;
import pl.marcinsalamandra.microbloggingapp.entity.AppUser;
import pl.marcinsalamandra.microbloggingapp.entity.Comment;
import pl.marcinsalamandra.microbloggingapp.entity.Picture;
import pl.marcinsalamandra.microbloggingapp.entity.Post;
import pl.marcinsalamandra.microbloggingapp.exception.NotFoundException;
import pl.marcinsalamandra.microbloggingapp.repository.CommentRepository;
import pl.marcinsalamandra.microbloggingapp.repository.LikeRepository;
import pl.marcinsalamandra.microbloggingapp.repository.PostRepository;

import java.io.IOException;

@Service
@RequiredArgsConstructor
@Transactional
public class PostService {

    private final PostRepository postRepository;
    private final LikeRepository likeRepository;
    private final CommentRepository commentRepository;
    private final AppUserService appUserService;
    private final PictureService pictureService;

    public void checkIfPostExists(Long postId) {
        if (!postRepository.existsById(postId)) {
            throw new NotFoundException("Post with id: " + postId + " not found");
        }
    }

    public Post getPost(Long postId) {
        return postRepository.findById(postId)
                .orElseThrow(() -> new NotFoundException("Post with id: " + postId + " not found"));
    }

    public Page<Post> getPosts(Pageable pageable) {
        return postRepository.findAll(pageable);
    }

    public Page<Post> getMostPopularPosts(Pageable pageable) {
        return postRepository.findMostPopularPosts(pageable);
    }

    public Page<Comment> getPostComments(Long postId, Pageable pageable) {
        checkIfPostExists(postId);

        return commentRepository.findPostComments(postId, pageable);
    }

    public Page<AppUser> getPostLikers(Long postId, Pageable pageable) {
        checkIfPostExists(postId);

        return appUserService.getPostLikers(postId, pageable);
    }

    public Picture getPostPicture(Long postId) {
        Post post = getPost(postId);

        if (!post.getHasPicture()) {
            throw new NotFoundException("Post picture not found");
        }

        return pictureService.getPictureByKey("post_" + postId);
    }

    public Post createPost(CreatePostDTO request) {
        return postRepository.save(new Post(
                appUserService.getUser(request.getPostedById()),
                request.getContent()
        ));
    }

    public void deletePost(Long postId) {
        checkIfPostExists(postId);

        commentRepository.deleteCommentsByPostId(postId);
        likeRepository.deleteLikesByPostId(postId);
        postRepository.deletePostById(postId);
    }

    public void updatePostPicture(Long postId, MultipartFile file) throws IOException {
        Post post = getPost(postId);

        pictureService.updatePicture("post_" + postId, file);
        post.setHasPicture(true);

        postRepository.save(post);
    }

    public void updatePost(Long postId, UpdatePostDTO request) {
        Post post = getPost(postId);
        post.setContent(request.getContent());

        if(post.getHasPicture() && !request.getHasPicture()) {
            pictureService.deletePicture("post_" + postId);
            post.setHasPicture(false);
        }

        postRepository.save(post);
    }
}

File: src/frontend/src/service/entity/PostService.ts
import {ContentType, HttpMethod, HttpService} from "../HttpService";
import {Page, Pageable} from "../../model/data/Page";
import {CreatePostRequest, Post, UpdatePostRequest} from "../../model/entity/Post";
import {User} from "../../model/entity/User";
import {Comment} from "../../model/entity/Comment";

export class PostService {

    public static getPost = (postId: string): Promise<Post> => {
        return HttpService.sendRequest<Post>(
            HttpMethod.GET,
            `/posts/${postId}`
        );
    }

    public static getPosts = (pageable: Pageable): Promise<Page<Post>> => {
        return HttpService.sendRequest<Page<Post>>(
            HttpMethod.GET,
            "/posts" + HttpService.addQueryParameters(pageable),
        );
    }

    public static getMostPopularPosts = (pageable: Pageable): Promise<Page<Post>> => {
        return HttpService.sendRequest<Page<Post>>(
            HttpMethod.GET,
            "/posts/popular" + HttpService.addQueryParameters(pageable),
        );
    }

    public static getPostLikers = (pageable: Pageable, postId: string): Promise<Page<User>> => {
        return HttpService.sendRequest<Page<User>>(
            HttpMethod.GET,
            `/posts/${postId}/likers` + HttpService.addQueryParameters(pageable),
        );
    }

    public static getPostComments = (pageable: Pageable, postId: string): Promise<Page<Comment>> => {
        return HttpService.sendRequest<Page<Comment>>(
            HttpMethod.GET,
            `/posts/${postId}/comments` + HttpService.addQueryParameters(pageable),
        );
    }

    public static createPost = (request: CreatePostRequest): Promise<Post> => {
        return HttpService.sendRequest<Post>(
            HttpMethod.POST,
            "/posts",
            JSON.stringify(request),
            ContentType.JSON
        );
    }

    public static deletePost = (postId: string): Promise<void> => {
        return HttpService.sendRequest<void>(
            HttpMethod.DELETE,
            `/posts/${postId}`
        );
    }

    public static updatePostPicture = (postId: string, picture: FormData): Promise<void> => {
        return HttpService.sendRequest<void>(
            HttpMethod.PUT,
            `/posts/${postId}/picture`,
            picture,
        );
    }

    public static updatePost = (postId: string, request: UpdatePostRequest): Promise<void> => {
        return HttpService.sendRequest<void>(
            HttpMethod.PUT,
            `/posts/${postId}`,
            JSON.stringify(request),
            ContentType.JSON
        );
    }
}